Very odd constructor call with transients

Hello! I've been trying to pin down a very odd issue I've been seeing with my larger code base and I've managed to make a trivial repro for it. It seems that calling `persistent` on a `vector_transient` is trying to instantiate a ctor call to my object using some transient impl type, but I've no idea why. This doesn't seem like expected behavior to me, so hopefully you can provide some insight.

```c++
#include <immer/vector.hpp>
#include <immer/vector_transient.hpp>
#include <immer/box.hpp>

struct object;

template <typename T>
struct is_valid_type
{ static bool constexpr value{ false }; };
template <>
struct is_valid_type<object>
{ static bool constexpr value{ true }; };

struct object
{
  using vector_type = immer::vector<immer::box<object>>;

  template <typename T>
  object(T &&)
  { static_assert(is_valid_type<T>::value, "invalid type"); }
};

int main()
{
  object::vector_type::transient_type t;
  // If this next line is commented out, the code compiles.
  (void)t.persistent();
}
```

Here's the error I'm getting from GCC 10.2.0 on Arch.

```
❯ c++ -I lib/immer main.cpp 
main.cpp: In instantiation of ‘object::object(T&&) [with T = immer::detail::rbts::rbtree<immer::box<object>, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap>, immer::refcount_policy, immer::spinlock_policy>, 5, 5>&]’:
lib/immer/immer/box.hpp:47:48:   required from ‘immer::box<T, MemoryPolicy>::holder::holder(Args&& ...) [with Args = {immer::detail::rbts::rbtree<immer::box<object, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::spinlock_policy, immer::no_transience_policy, false, true> >, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::spinlock_policy, immer::no_transience_policy, false, true>, 5, 5>&}; T = object; MemoryPolicy = immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap>, immer::refcount_policy, immer::spinlock_policy>]’
lib/immer/immer/detail/util.hpp:69:16:   required from ‘T* immer::detail::make(Args&& ...) [with Heap = immer::debug_size_heap<immer::cpp_heap>; T = immer::box<object>::holder; Args = {immer::detail::rbts::rbtree<immer::box<object, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::spinlock_policy, immer::no_transience_policy, false, true> >, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap, 1024>, immer::refcount_policy, immer::spinlock_policy, immer::no_transience_policy, false, true>, 5, 5>&}]’
lib/immer/immer/box.hpp:78:43:   required from ‘immer::box<T, MemoryPolicy>::box(Arg&&) [with Arg = immer::detail::rbts::rbtree<immer::box<object>, immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap>, immer::refcount_policy, immer::spinlock_policy>, 5, 5>&; Enable = void; T = object; MemoryPolicy = immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap>, immer::refcount_policy, immer::spinlock_policy>]’
lib/immer/immer/vector_transient.hpp:185:37:   required from ‘immer::vector_transient<T, MemoryPolicy, B, BL>::persistent_type immer::vector_transient<T, MemoryPolicy, B, BL>::persistent() & [with T = immer::box<object>; MemoryPolicy = immer::memory_policy<immer::free_list_heap_policy<immer::cpp_heap>, immer::refcount_policy, immer::spinlock_policy>; unsigned int B = 5; unsigned int BL = 5; immer::vector_transient<T, MemoryPolicy, B, BL>::persistent_type = immer::vector<immer::box<object> >]’
main.cpp:26:22:   required from here
main.cpp:20:37: error: static assertion failed: invalid type
   20 |   { static_assert(is_valid_type<T>::value, "invalid type"); }
      |                                     ^~~~~
```

The `immer::detail::rbts::rbtree` type *looks* like the `vector_transient::impl_t`, but I'm not sure. I don't think it should be trying to hit my object ctor with that though. :confused: 

Thank you for your time!

--- a/immer/flex_vector.hpp
+++ b/immer/flex_vector.hpp
@@ -500,14 +500,8 @@ public:
      * Returns an @a transient form of this container, an
      * `immer::flex_vector_transient`.
      */
-    IMMER_NODISCARD transient_type transient() const&
-    {
-        return transient_type{impl_};
-    }
-    IMMER_NODISCARD transient_type transient() &&
-    {
-        return transient_type{std::move(impl_)};
-    }
+    IMMER_NODISCARD transient_type transient() const& { return impl_; }
+    IMMER_NODISCARD transient_type transient() && { return std::move(impl_); }
 
     // Semi-private
     const impl_t& impl() const { return impl_; }

--- a/immer/flex_vector_transient.hpp
+++ b/immer/flex_vector_transient.hpp
@@ -231,12 +231,9 @@ public:
     IMMER_NODISCARD persistent_type persistent() &
     {
         this->owner_t::operator=(owner_t{});
-        return persistent_type{impl_};
+        return impl_;
     }
-    IMMER_NODISCARD persistent_type persistent() &&
-    {
-        return persistent_type{std::move(impl_)};
-    }
+    IMMER_NODISCARD persistent_type persistent() && { return std::move(impl_); }
 
 private:
     friend persistent_type;

--- a/immer/vector.hpp
+++ b/immer/vector.hpp
@@ -337,14 +337,8 @@ public:
      * Returns an @a transient form of this container, an
      * `immer::vector_transient`.
      */
-    IMMER_NODISCARD transient_type transient() const&
-    {
-        return transient_type{impl_};
-    }
-    IMMER_NODISCARD transient_type transient() &&
-    {
-        return transient_type{std::move(impl_)};
-    }
+    IMMER_NODISCARD transient_type transient() const& { return impl_; }
+    IMMER_NODISCARD transient_type transient() && { return std::move(impl_); }
 
     // Semi-private
     const impl_t& impl() const { return impl_; }

--- a/immer/vector_transient.hpp
+++ b/immer/vector_transient.hpp
@@ -182,12 +182,9 @@ public:
     IMMER_NODISCARD persistent_type persistent() &
     {
         this->owner_t::operator=(owner_t{});
-        return persistent_type{impl_};
+        return impl_;
     }
-    IMMER_NODISCARD persistent_type persistent() &&
-    {
-        return persistent_type{std::move(impl_)};
-    }
+    IMMER_NODISCARD persistent_type persistent() && { return std::move(impl_); }
 
 private:
     friend flex_t;

--- a/test/vector/issue-177.cpp
+++ b/test/vector/issue-177.cpp
@@ -1,0 +1,47 @@
+//
+// immer: immutable data structures for C++
+// Copyright (C) 2016, 2017, 2018 Juan Pedro Bolivar Puente
+//
+// This software is distributed under the Boost Software License, Version 1.0.
+// See accompanying file LICENSE or copy at http://boost.org/LICENSE_1_0.txt
+//
+
+// Thanks to @jeaye for reporting this issue
+// https://github.com/arximboldi/immer/issues/177
+
+#include <immer/box.hpp>
+#include <immer/vector.hpp>
+#include <immer/vector_transient.hpp>
+
+#include <catch.hpp>
+
+struct object;
+
+template <typename T>
+struct is_valid_type
+{
+    static bool constexpr value{false};
+};
+template <>
+struct is_valid_type<object>
+{
+    static bool constexpr value{true};
+};
+
+struct object
+{
+    using vector_type = immer::vector<immer::box<object>>;
+
+    template <typename T>
+    object(T&&)
+    {
+        static_assert(is_valid_type<T>::value, "invalid type");
+    }
+};
+
+TEST_CASE("strange bad constructor called")
+{
+    object::vector_type::transient_type t;
+    // If this next line is commented out, the code compiles.
+    (void) t.persistent();
+}

